### PR TITLE
dialects: (scf) add custom syntax for scf while

### PR DESCRIPTION
**Pull request by superlopuh**
Same as MLIR: https://mlir.llvm.org/docs/Dialects/SCFDialect/#scfwhile-scfwhileop

### CODE REVIEW

File: xdsl/dialects/scf.py [hunk 1]
```diff
@@ -81,6 +81,67 @@ def verify_(self):
...
+        function_type = parser.parse_function_type()
+
+        def resolve_argument(arg: parser.Argument, type: Attribute):
+            arg.type = type
+            return arg
+
+        block_args = tuple(
+            resolve_argument(block_arg, t)
+            for ((block_arg, _), t) in zip(
+                tuples, function_type.inputs.data, strict=True
```
Comment (math-fehr) [inline xdsl/dialects/scf.py:123]: You should return a Parsing Error if the function types and the inputs are not matching.

File: xdsl/dialects/scf.py [hunk 2]
```diff
@@ -81,6 +81,67 @@ def verify_(self):
...
+            )
+        )
+
+        arguments = tuple(
+            parser.resolve_operand(operand, t)
+            for ((_, operand), t) in zip(tuples, function_type.inputs.data, strict=True)
```
Comment (math-fehr) [inline xdsl/dialects/scf.py:129]: Same here

**Review by math-fehr — APPROVED** [review]
Just one error that I think should be catched, and looks good to me!

File: xdsl/dialects/scf.py (hunk 2 above)
Comment (superlopuh) [reply xdsl/dialects/scf.py:137]: it's actually the same value error we just iterate twice

**Review by webmiche — COMMENTED** [review]
Interop tests? :upside_down_face:

**Review by webmiche — APPROVED** [review]

**Pull request merged**